### PR TITLE
Add tls-assets secret sharding

### LINKED ISSUE #4353: TLS assets secret hits 1 MB etcd object size limit
**What happened?**
We're using service monitors to monitor https endpoints that need client tls certificates to retrieve metrics. After a certain amount of service monitors (250 ish) the operator is not able to save the tls-assets secret due to the k8s etcd object size limit.

**Did you expect to see something different?**
I'd expect it to keep working and scale beyond 250 service monitors

**How to reproduce it (as minimally and precisely as possible)**:
Create a big amount of service monitors with tls configs that include different client certs and drive the tls-assets secret size above 1048576 bytes.

**Environment**

* Prometheus Operator version:

    `quay.io/prometheus-operator/prometheus-operator:v0.47.0`
    
    

* Kubernetes version information:

    `Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.3", GitCommit:"ca643a4d1f7bfe34773c74f79527be4afd95bf39", GitTreeState:"clean", BuildDate:"2021-07-15T20:59:07Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"}`
    

* Kubernetes cluster kind:

    cluster-api

* Prometheus Operator Logs:

```
level=error ts=2021-10-29T10:17:49.181607935Z caller=klog.go:116 component=k8s_client_runtime func=ErrorDepth msg="Sync \"monitoring/k8s\" failed: creating tls asset secret failed: failed to create TLS assets secret for Prometheus k8s in namespace monitoring: Secret \"prometheus-k8s-tls-assets\" is invalid: data: Too long: must have at most 1048576 bytes"
```

### PR DESCRIPTION
**Pull request by gottwald**
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

k8s secrets have a default size limitation of 1048576 bytes. Once this
size gets exceeded the k8s API server will refuse to update the secret.
This introduces a sharding approach where all tls-assets are distributed
over shards of tls asset secrets and mounted into the prometheus pod
through volume projection.

Fixes #4353

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [x] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
- Add tls-assets secret sharding to support a high amount of tls assets
```

### CODE REVIEW

**Review by simonpasquier — COMMENTED** [review]
Very nice! We would need to implement the same logic for the Alertmanager controller too. What do you think about moving the code to github.com/prometheus-operator/prometheus-operator/pkg/operator package?

File: pkg/prometheus/operator.go [hunk 1]
```diff
@@ -1633,14 +1637,64 @@ func (c *Operator) createOrUpdateConfigurationSecret(ctx context.Context, p *mon
 	return k8sutil.CreateOrUpdateSecret(ctx, sClient, s)
 }
 
-func (c *Operator) createOrUpdateTLSAssetSecret(ctx context.Context, p *monitoringv1.Prometheus, store *assets.Store) error {
-	boolTrue := true
+func (c *Operator) createOrUpdateTLSAssetSecrets(ctx context.Context, p *monitoringv1.Prometheus, store *assets.Store) (assetShards int, err error) {
```
[the remark this answers is not in the document]
Comment (gottwald) [reply pkg/prometheus/operator.go:1640]: Just used it as a form of documentation w/o using naked returns. But I'm happy to drop it.

File: pkg/prometheus/operator.go [hunk 2]
```diff
@@ -1633,14 +1637,64 @@ func (c *Operator) createOrUpdateConfigurationSecret(ctx context.Context, p *mon
...
+	// k8s/etcd has an object limit of 1048576 bytes
+	// We use an upper limit of 950k for the data part to leave enough room for
+	// metadata and the rest of the object.
+	const maxSecretSizeBytes = 950_000
```
[the remark this answers is not in the document]
Comment (gottwald) [reply pkg/prometheus/operator.go:1644]: Nice!

**Comment by gottwald** [comment]
@simonpasquier moved it over to the operator package and tried to make it more generic. WDYT?

File: pkg/operator/sharded-secret.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+// shard does the in-memory sharding of the secret data.
+func (s *ShardedSecret) shard() []*v1.Secret {
+	// we need to iterate over the data in a stable order to avoid multiple
+	// re-shardings followed by secret updates just because the order of data
+	// items in the map changed and would now end up in another secret shard.
+	var keys []string
+	for k := range s.data {
+		keys = append(keys, k)
+	}
+	sort.Slice(keys, func(i, j int) bool { return keys[i] < keys[j] })
```
Comment (simonpasquier) [inline pkg/operator/sharded-secret.go:87]:
```suggestion
	sort.Strings(keys)
```

**Review by simonpasquier — COMMENTED** [review]
Thanks a lot! lgtm (minus one very small suggestion). If you don't have time to modify the alertmanager controller, I'm ok with merging your changes and we can take it from there.

**Comment by PhilipGough** [comment]
>  If you don't have time to modify the alertmanager controller, I'm ok with merging your changes and we can take it from there.

Perhaps it should be a follow up PR regardless?

**Comment by gottwald** [comment]
I'd favor making this a follow up. Not sure if I have time to pick it up. So maybe lets put in an issue to pick up later.

**Review by simonpasquier — APPROVED** [review]
LGTM. We'll keep #4353 until the same fix goes into the Alertmanager controller.

**Comment by simonpasquier** [comment]
Thanks a lot!

**Pull request merged**